packages/transactional-adapters/transactional-adapter-pg-promise/src/lib/transactional-adapter-pg-promise.ts

```diff
@@ -0,0 +1,35 @@
+import { TransactionalAdapter } from '@nestjs-cls/transactional';
+import { IDatabase, ITask } from 'pg-promise';
+
+export type PgPromiseDbOrTx = IDatabase<unknown> | ITask<unknown>;
```

Comment:
I am not sure about this type here, but I also don't know how to make it better... Ideally, I'd like `tx` to reference  a `Task` everywhere, but this is impossible without wrapping a callback in `db.task`, but the `getFallbackInstance` doesn't support this (and can't support it, really).

Given that, I am inclined in typing it as `IDatabase` (`ITask` just adds the `ctx`, `batch`, `sequence` and `page` methods) and when the user wants to use a method on the Task interface, they can wrap their db call with `txHost.tx.task(t =>  t.batch(...))`, which should re-use the existing client in case we are in a transaction or get a new one form the pool for the duration of the task.

Reply:
You correctly understood that my intent was to handle nested transactions. But [batch](https://vitaly-t.github.io/pg-promise/Task.html#batch) is deprecated – actually I think it's a leftover from the times when async/await weren't widely used.

I'd rather keep it this way unless someone comes with a better idea?

Reply:
Alright, will type it as `IDatabase<unknown>`, then.

Reply:
In the end I had to type it `ITask<unknown>` instead. This is because [this callback](https://github.com/sam-artuso/nestjs-cls/blob/feat/transactional-adapter-pg-promise/packages/transactional-adapters/transactional-adapter-pg-promise/src/lib/transactional-adapter-pg-promise.ts#L32-L35) must have this signature `(t: ITask<unknown>) => any`.

Reply:
But that is unsafe ftom the typing perspective :/ Accessing `ctx` on `txHost.tx` or other properties that should exist on `Task` outside of transaction will get you undefined and a runtime failure. I think it would be safer to type both the `TConnection` _and_ the `TTx` as `IDatabase` instead.

Reply:
I can see the issue you're talking about now.

I'm not sure how to proceeed from here.

```ts
            return pgPromiseDbInstance.tx(options ?? {}, (tx) => {
                                                       // ^ ITask<unknown>
                setClient(tx);
                       // ^ Argument of type 'ITask<unknown>' is not assignable to parameter of type 'IDatabase<unknown, IClient>'.
                return fn();
            });
```

I can't change the signature of `setClient()` either, because that would break `optionsFactory()`. Help needed!

Reply:
Hm, at that point I would probably just use a type assertion `as IDatabase<unknown>`. An `ITask` is a subtype of `IDatabase` so it should generally work. Internal adapter code can't be often typed 100% correctly, but we should at least aim for the most usable and correct public API.

Reply:
Uhm, actually that's not a good idea because it would make accessing the task's `ctx` impossible. I'm at a loss.

Best to do `as unknown as IDatabase<unknown>`, then?

Reply:
> it would make accessing the task's ctx impossible
That's not entirely wrong then. The transaction management should be left to `@nestjs-cls/transactional` anyway. So as long as `IBaseProtocol` has all the methods to prepare and run queries and is properly typed, it should be fine. An explicit task can be always created with `txHost.tx.task(task => { ... })`, right?

Reply:
But I'm not opposed to typing it as `Database`, since reaching into `IBaseProtocol` might not be safe because it's not a public documented API and the documentation says that [Task extends Database](http://vitaly-t.github.io/pg-promise/Task.html).

Reply:
Ok, all done. Are you happy with the PR overall?